File: smalivm/src/test/java/org/cf/smalivm/dex/ClassManagerTest.java
package org.cf.smalivm.dex;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import org.cf.smalivm.VMTester;
import org.cf.smalivm.type.ClassManager;
import org.cf.smalivm.type.ClassManagerFactory;
import org.cf.smalivm.type.VirtualArray;
import org.cf.smalivm.type.VirtualClass;
import org.cf.smalivm.type.VirtualPrimitive;
import org.cf.smalivm.type.VirtualType;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

public class ClassManagerTest {

    private static ClassManager classManager;

    @BeforeAll
    public static void getClassManager() throws IOException {
        classManager = new ClassManagerFactory().build(VMTester.TEST_CLASS_PATH);
    }

    @Test
    public void canGetVirtualClassFromString() {
        VirtualClass virtualClass = classManager.getVirtualClass(CommonTypes.STRING);
        assertEquals(CommonTypes.STRING, virtualClass.getName());
    }

    @Test
    public void canGetVirtualClassFromClass() {
        VirtualClass virtualClass = classManager.getVirtualClass(String.class);
        assertEquals(CommonTypes.STRING, virtualClass.getName());
    }

    @Test
    public void canGetVirtualPrimitiveFromString() {
        VirtualType virtualType = classManager.getVirtualType(CommonTypes.INTEGER);
        assertEquals(CommonTypes.INTEGER, virtualType.getName());
        assertTrue(virtualType instanceof VirtualPrimitive);
    }

    @Test
    public void canGetVirtualPrimitiveFromClass() {
        VirtualType virtualType = classManager.getVirtualType(CommonTypes.INTEGER);
        assertEquals(CommonTypes.INTEGER, virtualType.getName());
        assertTrue(virtualType instanceof VirtualPrimitive);
    }

    @Test
    public void canGetVirtualArrayFromString() {
        String typeSignature = "[" + CommonTypes.STRING;
        VirtualType virtualType = classManager.getVirtualType(typeSignature);
        assertEquals(typeSignature, virtualType.getName());
        assertTrue(virtualType instanceof VirtualArray);
    }

    @Test
    public void canGetVirtualArrayFromClass() {
        VirtualType virtualType = classManager.getVirtualType(String[].class);
        assertEquals("[" + CommonTypes.STRING, virtualType.getName());
        assertTrue(virtualType instanceof VirtualArray);
    }
}

File: smalivm/src/test/java/org/cf/smalivm/dex/ClassManagerTest.kt
package org.cf.smalivm.dex

import org.cf.smalivm.Tester
import org.cf.smalivm.type.ClassManager
import org.cf.smalivm.type.ClassManagerFactory
import org.cf.smalivm.type.VirtualArray
import org.cf.smalivm.type.VirtualPrimitive
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.io.IOException

class ClassManagerTest {
    @Test
    fun `Can get virtual class from string`() {
        val virtualClass = classManager.getVirtualClass(CommonTypes.STRING)
        Assertions.assertEquals(CommonTypes.STRING, virtualClass.name)
    }

    @Test
    fun `Can get virtual class from class`() {
        val virtualClass = classManager.getVirtualClass(String::class.java)
        Assertions.assertEquals(CommonTypes.STRING, virtualClass.name)
    }

    @Test
    fun `Can get virtual primitive from string`() {
        val virtualType = classManager.getVirtualType(CommonTypes.INTEGER)
        Assertions.assertEquals(CommonTypes.INTEGER, virtualType.name)
        Assertions.assertTrue(virtualType is VirtualPrimitive)
    }

    @Test
    fun `Can get virtual primitive from class`() {
        val virtualType = classManager.getVirtualType(CommonTypes.INTEGER)
        Assertions.assertEquals(CommonTypes.INTEGER, virtualType.name)
        Assertions.assertTrue(virtualType is VirtualPrimitive)
    }

    @Test
    fun `Can get virtual array from string`() {
        val typeSignature = "[" + CommonTypes.STRING
        val virtualType = classManager.getVirtualType(typeSignature)
        Assertions.assertEquals(typeSignature, virtualType.name)
        Assertions.assertTrue(virtualType is VirtualArray)
    }

    @Test
    fun `Can get virtual array from class`() {
        val virtualType = classManager.getVirtualType(Array<String>::class.java)
        Assertions.assertEquals("[" + CommonTypes.STRING, virtualType.name)
        Assertions.assertTrue(virtualType is VirtualArray)
    }

    companion object {
        private lateinit var classManager: ClassManager

        @BeforeAll
        @JvmStatic
        @Throws(IOException::class)
        fun getClassManager() {
            classManager = ClassManagerFactory().build(Tester.TEST_CLASS_PATH)
        }
    }
}
